### ISSUE TITLE
A few Destroy Issues

### ISSUE THREAD

**Issue by Fox-McCloud** [tgstation/-tg-station#12143]
Just a few Destroy related issues:

- De-referencing hotspots is handled in kill() and Destroy, when everything should be located in Destroy---likewise, hotspot Destroy's don't call parent: https://github.com/tgstation/-tg-station/blob/a77b9c75be776e06990ba4b5ccc68e9656ed0a72/code/LINDA/LINDA_fire.dm#L144
 - This is also directly calling PlaceInPool(), which shouldn't be done; that should be handled in qdel hints. Related, Kill() is also used to semi-qdel() hotspots instead of qdel()'ing hotspots directly.

- De-referencing of bots is handled in explode() and Destroy: this results in bots attempting to be processed that have been qdel'd directly: https://github.com/tgstation/-tg-station/blob/15e50de0feb35850e348919d8f20cb63777fafc7/code/game/machinery/bots/bots.dm#L113

- Emitters beam's destroy doesn't call parent: https://github.com/tgstation/-tg-station/blob/009a53271ec9eb72ff0bdc181cb2dcc31ba6db62/code/modules/projectiles/projectile/beams.dm#L66

- slaughter demon's Destroy doesn't call parent.
 https://github.com/tgstation/-tg-station/blob/01c362f06904f6928a9d7a08f0314e72e128c9c3/code/modules/mob/living/bloodcrawl.dm#L29

- On screen alerts' Destroy doesn't call parent: https://github.com/tgstation/-tg-station/blob/9fbabcc3052f61eb850dc2b5786f90a413903162/code/_onclick/hud/alert.dm#L292 

- Virtual speakers are placed directly in a pool instead of being qdel()'d and having a qdel hint to place them in a pool---place in pool shouldn't ever be called directly, AFAIK: https://github.com/tgstation/-tg-station/blob/009a53271ec9eb72ff0bdc181cb2dcc31ba6db62/code/game/machinery/telecomms/broadcasting.dm#L172

**Comment by duncathan** [comment]
I love when people find this shit thanks fox
QDEL is my favorite system besides pipenets

**Comment by Fox-McCloud** [comment]
@duncathan 

Not a problem; I enjoy finding them---proper garbage collection is integral to SS13 performance, and the better references are cleared (and there are no issues with dangling references or oddities like processing non-existent bots) the better it is for coders and players alike.

**Comment by RemieRichards** [comment]
Emitter beams doing that is because they were the first thing Pooled, meaning they were one of the first things with a custom ```Destroy()``` that did something other than dereferencing.
There is NOTHING wrong with calling ```PlaceInPool()``` directly, it's just not advised anymore as you can just return ```QDEL_HINT_PUTINPOOL```

**Comment by duncathan** [comment]
@RemieRichards in #12153 I added a comment saying that most of the time `QDEL_HINT_PUTINPOOL` should be used over `PlaceInPool()`. Is that accurate?

**Comment by RemieRichards** [comment]
@duncathan yes, that's how we do things now, but the vibe I was getting from Fox and a few others was that they felt calling PlaceInPool() directly was Voodoo that shouldn't be done for fear of summoning Satan, where as it's just a formatting thing really.

tl;dr yes, but it's not a safety thing or anything.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

I know you specifically designed it so that you can place it in the pool directly, but, I was under the understanding this is still a "no-no" and definitely a consistency issue---ie: it's like calling a proc like "killthething" which deferences that thing then qdel's it at the end---with very little dereferencing done in the actual destroying---ie: it works, and you could use it that way, but it's inconsistent.

**Comment by RemieRichards** [comment]
@Fox-McCloud yes, it's just a consistency issue, but as I said, I was getting the vibe from you and others that you thought it would break the universe in two if you even tried to do it.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

Nah, if I came across that way, sorry.
